File: packages/standalone-python/src/trafilaturacore/_images.py
# SPDX-License-Identifier: Apache-2.0
"""Offline image source and alt-text handling before the native sanitizer."""

from __future__ import annotations

import re
from dataclasses import dataclass
from urllib.parse import urljoin, urlsplit

from lxml.html import HtmlElement

from ._result import Message
from ._utils import trim


@dataclass
class SrcsetCandidate:
    """One candidate with optional width and density descriptors."""

    url: str
    width: float | None = None
    density: float | None = None
    valid_descriptor: bool = True


def _space(char: str) -> bool:
    """Use HTML's ASCII whitespace when scanning srcset."""
    return bool(char) and char in " \t\n\f\r"


def parse_srcset(value: str) -> list[SrcsetCandidate]:
    """Translate the primary URL-first scan, keeping data-URI commas intact."""
    result = []
    position = 0
    while position < len(value):
        while position < len(value) and (_space(value[position]) or value[position] == ","):
            position += 1
        start = position
        while position < len(value) and not _space(value[position]):
            position += 1
        end = position
        while end > start and value[end - 1] == ",":
            end -= 1
        candidate = SrcsetCandidate(value[start:end])
        descriptors = 0
        if end == position:
            while position < len(value):
                while position < len(value) and _space(value[position]):
                    position += 1
                if position == len(value):
                    break
                if value[position] == ",":
                    position += 1
                    break
                start = position
                while (
                    position < len(value) and not _space(value[position]) and value[position] != ","
                ):
                    position += 1
                descriptors += 1
                descriptor = value[start:position]
                match = re.fullmatch(
                    r"([0-9]+(?:\.[0-9]+)?)([wx])", descriptor, re.IGNORECASE | re.ASCII
                )
                if descriptors > 1 or match is None or descriptor[-1:] not in {"w", "x"}:
                    candidate.valid_descriptor = False
                if match:
                    size = float(match[1])
                    if match[2].lower() == "w":
                        candidate.width = size
                    else:
                        candidate.density = size
        if candidate.url and not candidate.url.lower().startswith("data:"):
            result.append(candidate)
    return result


def _largest(candidates: list[SrcsetCandidate]) -> str | None:
    """Prefer widths, then density, preserving the first equal-ranked candidate."""
    use_width = any(candidate.width is not None for candidate in candidates)
    best = None
    rank = float("-inf")
    for candidate in candidates:
        if use_width and candidate.width is None:
            continue
        value = candidate.width if use_width else candidate.density
        value = (0 if use_width else 1) if value is None else value
        if best is None or value > rank:
            best, rank = candidate, value
    return best.url if best is not None else None


def _usable(value: str | None) -> str | None:
    """Ignore missing, blank and data-URI source candidates."""
    normalized = value.strip() if value is not None else ""
    return normalized if normalized and not normalized.lower().startswith("data:") else None


def _source(image: HtmlElement) -> str | None:
    """Preserve lazy/canonical source precedence before considering srcset."""
    for name in ("data-src", "src"):
        candidate = _usable(image.get(name))
        if candidate is not None:
            return candidate
    for name, value in image.attrib.items():
        if name.startswith("data-src") and name not in {"data-src", "data-srcset"}:
            candidate = _usable(value)
            if candidate is not None:
                return candidate
    own = image.get("srcset")
    if own is None:
        own = image.get("data-srcset")
    candidates = parse_srcset(own) if own is not None else []
    if not candidates:
        picture = next(image.iterancestors("picture"), None)
        if picture is not None:
            for source in picture.iter("source"):
                raw = source.get("srcset")
                if raw is None:
                    raw = source.get("data-srcset")
                if raw is not None:
                    candidates.extend(parse_srcset(raw))
    return _largest(candidates)


def _alt(
    image: HtmlElement, figures: dict[HtmlElement, list[HtmlElement]], ids: dict[str, HtmlElement]
) -> tuple[str | None, bool]:
    """Apply explicit alt, caption, ARIA and title fallbacks in primary order."""
    explicit = image.get("alt")
    if explicit is not None:
        value = trim(explicit)
        return value or None, not value
    figure = next(image.iterancestors("figure"), None)
    if figure is not None and len(figures.get(figure, [])) == 1:
        # Native captions historically joined all caption text segments.
        caption = trim(
            " ".join(text for node in figure.iter("figcaption") for text in node.itertext())
        )
        if caption:
            return caption, False
    label = trim(image.get("aria-label", ""))
    if label:
        return label, False
    descriptions = [
        trim("".join(ids[key].itertext()))
        for key in image.get("aria-labelledby", "").split()
        if key in ids
    ]
    labelled = " ".join(filter(None, descriptions))
    if labelled:
        return labelled, False
    return trim(image.get("title", "")) or None, False


def apply_images(
    document: HtmlElement, mode: str, base: str | None, messages: list[Message]
) -> None:
    """Transform images in place while retaining native fragment/URL semantics."""
    if mode not in {"alt-text", "resolved-url"}:
        return
    ids = {node.get("id"): node for node in document.iter() if node.get("id") is not None}
    figures = {figure: list(figure.iter("img")) for figure in document.iter("figure")}
    unresolved = invalid = False
    for image in list(document.iter("img")):
        if mode == "alt-text":
            alt, decorative = _alt(image, figures, ids)
            if decorative:
                if image.getparent() is not None:
                    image.drop_tree()
                continue
            image.attrib.clear()
            if alt:
                image.set("alt", alt)
            continue
        candidate = _source(image)
        for name in list(image.attrib):
            if name not in {"src", "alt", "title", "width", "height"}:
                del image.attrib[name]
        if candidate is None:
            continue
        try:
            resolved = urljoin(base or "", candidate)
            parsed = urlsplit(resolved)
        except ValueError:
            image.attrib.pop("src", None)
            invalid = True
        else:
            image.set("src", resolved)
            unresolved |= not base and not bool(parsed.scheme)
    for tag in ("source", "picture"):
        for element in list(document.iter(tag)):
            if element.getparent() is not None:
                element.drop_tag()
    for figure, previous in figures.items():
        if previous and next(figure.iter("img"), None) is None and figure.getparent() is not None:
            figure.drop_tree()
    if unresolved:
        messages.append(Message("warning", "relative image URL left unchanged without a base URL"))
    if invalid:
        messages.append(Message("warning", "invalid image URL removed"))

File: packages/standalone/src/cleaning/images.ts
// SPDX-License-Identifier: Apache-2.0
// First-party pre-sanitize image policy. Extraction modes see only attributes retained
// by the upstream extractor; whole-document mode can also use picture, ARIA and captions.
// URL resolution is lexical only. Every result still crosses the sanitizer/security floor.

import {
  countThroughTemplates,
  type HDocument,
  type HElement,
  parseDocument,
  querySelectorAllThroughTemplates,
  trim,
} from '../metadata/dom.js';
import type { Message } from '../types.js';
import { isHtmlDocument } from './normalize.js';

export interface ImagePassOptions {
  mode: 'alt-text' | 'resolved-url';
  baseUrl?: string;
}
export interface ImagePassResult {
  html: string;
  messages: Message[];
}
interface SrcsetCandidate {
  url: string;
  width?: number;
  density?: number;
}

export function readBaseHref(rawHtml: string): string | undefined {
  const value = parseDocument(rawHtml).querySelector('base[href]')?.getAttribute('href')?.trim();
  if (!value) return undefined;
  try {
    new URL(value);
    return value;
  } catch {
    return undefined;
  }
}

function space(value: string | undefined): boolean {
  return value !== undefined && ' \t\n\f\r'.includes(value);
}

/** Scan URLs before descriptors, so the comma in a data URI is not a candidate separator. */
export function parseSrcset(value: string): SrcsetCandidate[] {
  const result: SrcsetCandidate[] = [];
  let position = 0;
  while (position < value.length) {
    while (space(value[position]) || value[position] === ',') position += 1;
    const start = position;
    while (position < value.length && !space(value[position])) position += 1;
    let end = position;
    while (end > start && value[end - 1] === ',') end -= 1;
    const candidate: SrcsetCandidate = { url: value.slice(start, end) };
    if (end === position) {
      while (position < value.length) {
        while (space(value[position])) position += 1;
        if (value[position] === ',') {
          position += 1;
          break;
        }
        if (position === value.length) break;
        const from = position;
        while (position < value.length && !space(value[position]) && value[position] !== ',')
          position += 1;
        const descriptor = /^(\d+(?:\.\d+)?)([wx])$/i.exec(value.slice(from, position));
        if (descriptor !== null) {
          const size = Number(descriptor[1]);
          if (descriptor[2]?.toLowerCase() === 'w') candidate.width = size;
          else candidate.density = size;
        }
      }
    }
    if (candidate.url && !candidate.url.toLowerCase().startsWith('data:')) result.push(candidate);
  }
  return result;
}

function largest(candidates: readonly SrcsetCandidate[]): string | undefined {
  const useWidth = candidates.some((candidate) => candidate.width !== undefined);
  let best: SrcsetCandidate | undefined;
  let rank = -Infinity;
  for (const candidate of candidates) {
    if (useWidth && candidate.width === undefined) continue;
    const value = useWidth ? (candidate.width ?? 0) : (candidate.density ?? 1);
    if (best === undefined || value > rank) {
      best = candidate;
      rank = value;
    }
  }
  return best?.url;
}

function usable(value: string | null): string | undefined {
  const url = value?.trim();
  return url && !url.toLowerCase().startsWith('data:') ? url : undefined;
}

function source(image: HElement, picture: readonly SrcsetCandidate[]): string | undefined {
  for (const name of ['data-src', 'src']) {
    const value = usable(image.getAttribute(name));
    if (value !== undefined) return value;
  }
  for (const attribute of image.attributes) {
    const name = attribute.name.toLowerCase();
    if (name.startsWith('data-src') && name !== 'data-src' && name !== 'data-srcset') {
      const value = usable(attribute.value);
      if (value !== undefined) return value;
    }
  }
  const set = image.getAttribute('srcset') ?? image.getAttribute('data-srcset');
  const own = set === null ? [] : parseSrcset(set);
  return largest(own.length ? own : picture);
}

function pictures(
  document: HDocument,
  mode: ImagePassOptions['mode'],
): Map<HElement, SrcsetCandidate[]> {
  const sources = new Map<HElement, SrcsetCandidate[]>();
  for (const picture of querySelectorAllThroughTemplates(document, 'picture')) {
    const image = picture.querySelector('img');
    if (image === null) {
      picture.remove();
      continue;
    }
    if (mode === 'resolved-url') {
      const candidates: SrcsetCandidate[] = [];
      for (const source of picture.querySelectorAll('source')) {
        const raw = source.getAttribute('srcset') ?? source.getAttribute('data-srcset');
        if (raw !== null) for (const candidate of parseSrcset(raw)) candidates.push(candidate);
      }
      if (candidates.length) sources.set(image, candidates);
    }
    picture.replaceWith(image);
  }
  return sources;
}

function alt(
  image: HElement,
  document: HDocument,
): { text: string | undefined; decorative: boolean } {
  const explicit = image.getAttribute('alt');
  if (explicit !== null) {
    const text = trim(explicit);
    return { text: text || undefined, decorative: !text };
  }
  const figure = image.closest('figure');
  if (figure !== null && countThroughTemplates(figure, 'img', 1) === 1) {
    const caption = trim(figure.querySelector('figcaption')?.textContent ?? '');
    if (caption) return { text: caption, decorative: false };
  }
  const label = trim(image.getAttribute('aria-label') ?? '');
  if (label) return { text: label, decorative: false };
  const descriptions: string[] = [];
  for (const id of (image.getAttribute('aria-labelledby') ?? '').split(/\s+/)) {
    if (!id || id.includes('"') || id.includes('\\')) continue;
    const text = trim(document.querySelector(`[id="${id}"]`)?.textContent ?? '');
    if (text) descriptions.push(text);
  }
  const labelled = descriptions.join(' ');
  if (labelled) return { text: labelled, decorative: false };
  return { text: trim(image.getAttribute('title') ?? '') || undefined, decorative: false };
}

const KEEP = new Set(['src', 'alt', 'title', 'width', 'height']);

export function applyImageHandling(html: string, options: ImagePassOptions): ImagePassResult {
  const document = parseDocument(html);
  const fallback = pictures(document, options.mode);
  const figures = querySelectorAllThroughTemplates(document, 'figure').filter(
    (figure) => countThroughTemplates(figure, 'img', 0) > 0,
  );
  let unresolved = 0;
  for (const image of querySelectorAllThroughTemplates(document, 'img')) {
    if (options.mode === 'alt-text') {
      const derived = alt(image, document);
      if (derived.decorative) image.remove();
      else {
        const placeholder = document.createElement('img');
        if (derived.text !== undefined) placeholder.setAttribute('alt', derived.text);
        image.replaceWith(placeholder);
      }
      continue;
    }
    const candidate = source(image, fallback.get(image) ?? []);
    for (const attribute of [...image.attributes]) {
      if (!KEEP.has(attribute.name.toLowerCase())) image.removeAttribute(attribute.name);
    }
    if (candidate === undefined) continue;
    let resolved = candidate;
    try {
      resolved = new URL(candidate, options.baseUrl).toString();
    } catch {
      unresolved += 1;
    }
    image.setAttribute('src', resolved);
  }
  for (const figure of figures) {
    if (countThroughTemplates(figure, 'img', 0) === 0) figure.remove();
  }
  const messages: Message[] = unresolved
    ? [
        {
          type: 'warning',
          text:
            "imageHandling 'resolved-url': no base URL (no url option, no <base href>) — " +
            unresolved +
            ' relative image URL(s) left as authored',
        },
      ]
    : [];
  return {
    html: isHtmlDocument(html) ? String(document) : (document.body?.innerHTML ?? ''),
    messages,
  };
}
